merlion/models/deep_base.py

```diff
@@ -0,0 +1,228 @@
+        :param num_epochs: Total number of epochs for training.
+        :param optimizer: The optimizer for learning the parameters of the deep learning models. The value of optimizer
+            can be ``Adam``, ``AdamW``, ``SGD``, ``Adagrad``, ``RMSprop``.
+        :param loss_fn: Loss function for optimizing deep learning models. The value of loss_fn can be `mse` for l2 loss,
+            `l1` for l1 loss, `huber` for huber loss.
+        :param clip_gradient: Clipping gradient norm of model parameters before updating. If `clip_gradient is None`,
+            then the gradient will not be clipped.
+        :param use_gpu: Whether to use gpu for training deep models. If ``use_gpu = True`` while thre is no GPU device,
+            the model will use CPU for training instead.
+        :param ts_encoding: whether the timestamp should be encoded to a float vector, which can be used
+            for training deep learning based time series models; if ``None``, the timestamp is not encoded.
+            If not ``None``, it represents the frequency for time features encoding options:[s:secondly, t:minutely, h:hourly,
+            d:daily, b:business days, w:weekly, m:monthly]
+        :param lr: Learning rate for optimizing deep learning models.
+        :param weight_decay: Weight decay (L2 penalty) (default: 0)
+        :param valid_fraction: Fraction of validation set splitted from training data
+        :param early_stop_patience: Number of epochs with no improvement after which training will be stopped for
+            early stopping function. If `early_stop_patience = None`, the training process will not stop early.
+        """
+        super().__init__(**kwargs)
+
+    @property
+    def optimizer(self) -> Optimizer:
+        return self._optimizer
+
+    @optimizer.setter
+    def optimizer(self, optimizer: Union[str, Optimizer]):
+        if isinstance(optimizer, str):
+            valid = set(Optimizer.__members__.keys())
+            if optimizer not in valid:
+                raise KeyError(f"{optimizer} is not a valid optimizer that supported. Valid optimizers are: {valid}")
+            optimizer = Optimizer[optimizer]
+        self._optimizer = optimizer
+
+    @property
+    def loss_fn(self) -> LossFunction:
+        return self._loss_fn
+
+    @loss_fn.setter
+    def loss_fn(self, loss_fn: Union[str, LossFunction]):
+        if isinstance(loss_fn, str):
+            valid = set(LossFunction.__members__.keys())
+            if loss_fn not in valid:
+                raise KeyError(f"{loss_fn} is not a valid loss that supported. Valid optimizers are: {valid}")
+            loss_fn = LossFunction[loss_fn]
+        self._loss_fn = loss_fn
+
+
+class TorchModel(nn.Module):
+    """
+    Abstract base class for Pytorch deep learning models
+    """
+
+    def __init__(self, config: DeepConfig):
+        super(TorchModel, self).__init__()
+        self.config = config
+
+    @abstractmethod
+    def forward(self, past, *args, **kwargs):
```

Comment:
Maybe you can set the signature to `past, past_timestamp, future_timestamp, **kwargs`, since all of the models are like this?